### PR TITLE
Bug 1707617 - Implement `init` and `set_upload_enabled` reasons for deletion-request ping

### PR DESCRIPTION
**Pull request by AmrDhaliwal**
### Pull Request checklist ###

- [ ] **Quality**: Make sure this PR builds and runs cleanly.
  - Inside the `glean/` folder, run:
    - `npm run test` Runs _all_ tests
    - `npm run lint` Runs _all_ linters
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry to `CHANGELOG.md` or an explanation of why it does not need one
- [ ] **Documentation**: This PR includes documentation changes, an explanation of why it does not need that or a follow-up bug has been filed to do that work
  - Documentation should be added to [The Glean Book](https://mozilla.github.io/glean/book/index.html) when making changes to Glean's user facing API
    - When changes to the Glean Book are necessary, link to the corresponding PR on the [`mozilla/glean`](https://github.com/mozilla/glean) repository
  - Documentation should be added to [the Glean.js developer documentation](https://github.com/mozilla/glean.js/tree/main/docs) when making internal code changes

### CODE REVIEW

**Comment by AmrDhaliwal** [comment]
@brizental the CircleCI job got put on hold will be back tomorrow to finish changes for any failed tests and to add to the changelog.

**Review by brizental — CHANGES_REQUESTED** [review]
Hey @AmrDhaliwal this looks great. Thank you so much for the contribution.

There a few things to do before we can merge it:

1. Please update the `pings.yaml` deletion-request definition to contain the reasons. You can copy what we have over in mozilla/glean: https://github.com/mozilla/glean/blob/1642abda89bae9b27b74f9b59f7ef1da2e6b65e9/glean-core/pings.yaml#L105-L131
2. Please add a test to verify that the resons are the expected.

Thanks!

**Comment by AmrDhaliwal** [comment]
Hi @brizental,

I updated the pings.yaml but, am unsure of where to add the test for the setUploadEnabled() and OnUploadDisabled() functions. I was hoping you could steer me in the right direction?

File: glean/src/core/glean.ts [hunk 1]
```diff
@@ -57,15 +57,23 @@ namespace Glean {
    * A deletion_request ping is sent, all pending metrics, events and queued
    * pings are cleared, and the client_id is set to KNOWN_CLIENT_ID.
    * Afterward, the upload_enabled flag is set to false.
+   *
+   * @param at_init When true contains the "at_init" reason
```
Comment (brizental) [inline glean/src/core/glean.ts:61]:
```suggestion
   * @param at_init Whether or not upload has been disabled during initialization.
```

**Review by brizental — CHANGES_REQUESTED** [review]
> I was hoping you could steer me in the right direction?

Sure. We need two checks:

1. Check that the ping contains the expected reason when sent during initialization
2. Check that the ping contains the expected reason when sent through `setUploadEnabled`

We have tests already for both deletion-requests sent on init and when upload status is toggled. So we can add an extra check in each test. [This](https://github.com/mozilla/glean.js/blob/main/glean/tests/unit/core/glean.spec.ts#L316-L335) is the test that checks the deletion-request sent on init and [this](https://github.com/mozilla/glean.js/blob/main/glean/tests/unit/core/glean.spec.ts#L276-L291) is the test that checks the deletion-request sent when upload is toggled.

Let me knos if anything is still unclear. Happy to clarify.

**Comment by AmrDhaliwal** [comment]
Hi @brizental sorry, I think I'm still a little confused. Should I add the check, in the tests or create new tests for the checks?

**Comment by AmrDhaliwal** [comment]
@brizental I was hoping you could let me know if I'm heading in the right direction. Should I run the testInititalizeGlean function and see if the pingBody contains the reason?

**Comment by brizental** [comment]
> @brizental I was hoping you could let me know if I'm heading in the right direction. Should I run the testInititalizeGlean function and see if the pingBody contains the reason?

Yeap that is correct.

**Comment by AmrDhaliwal** [comment]
@brizental Could you look over this? I seem to be getting an unsafe member access of an "any" value and am
unsure of the best way to solve it.

**Comment by brizental** [comment]
> @brizental Could you look over this? I seem to be getting an unsafe member access of an "any" value and am
unsure of the best way to solve it.

Sure, I assume the error is happening on [this line](https://github.com/mozilla/glean.js/pull/1154/files#diff-d45b4e0a01e3422fefd2d145ebb8340125f40dd97f34b76deada19fb6635e446R341). Is that correct? If so, that error is probably happening because `pingBody` has type JSONObject and so TypeScript has no way to know what is the type of `ping_info`. Since these are tests, we are fine coercing values like: `(info as JSONObject).reason`.

**Comment by AmrDhaliwal** [comment]
@brizental good you look at this when have a moment

**Review by brizental — APPROVED** [review]
Thanks a lot @AmrDhaliwal this is looking great! Would you mind adding this change to the CHANGELOG.md file? Otherwise, good to go.

**Comment by brizental** [comment]
Hey @AmrDhaliwal would you mind also rebasing this so we can merge? Thanks a lot!

**Comment by brizental** [comment]
Hey Amrit I took the liberty to fix the merge issues here so I can merge. I am cutting a new release of glean.js today and this is an important change to include in it. Thank you again for the hard work here, much appreciated!

**Pull request merged**